File: 954-maximum-sum-circular-subarray/maximum-sum-circular-subarray.c
int kadane(int arr[], int n) {
    int max_sum = INT_MIN;
    int current_sum = 0;

    for (int i = 0; i < n; i++) {
        current_sum = (arr[i] > current_sum + arr[i]) ? arr[i] : current_sum + arr[i];
        max_sum = (max_sum > current_sum) ? max_sum : current_sum;
    }

    return max_sum;
}

int maxSubarraySumCircular(int* nums, int numsSize) {
    int max_inner_sum = kadane(nums, numsSize);

    int total_sum = 0;
    int inverted_nums[numsSize];

    for (int i = 0; i < numsSize; i++) {
        total_sum += nums[i];
        inverted_nums[i] = -nums[i];
    }

    int max_wrap_sum = total_sum + kadane(inverted_nums, numsSize);

    return (max_wrap_sum != 0) ? ((max_inner_sum > max_wrap_sum) ? max_inner_sum : max_wrap_sum) : max_inner_sum;
}

File: 954-maximum-sum-circular-subarray/maximum-sum-circular-subarray.java
class Solution {
    public int maxSubarraySumCircular(int[] nums) {
        int n = nums.length;

        int maxInnerSum = kadane(nums);

        int totalSum = 0;
        for (int num : nums) {
            totalSum += num;
        }

        int[] invertedNums = new int[n];
        for (int i = 0; i < n; i++) {
            invertedNums[i] = -nums[i];
        }

        int maxWrapSum = totalSum + kadane(invertedNums);

        return Math.max(maxInnerSum, (maxWrapSum == 0) ? maxInnerSum : maxWrapSum);
    }

    private int kadane(int[] arr) {
        int maxSum = Integer.MIN_VALUE;
        int currentSum = 0;

        for (int num : arr) {
            currentSum = Math.max(num, currentSum + num);
            maxSum = Math.max(maxSum, currentSum);
        }

        return maxSum;
    }
}